src/Widgets/FirmwareUpdateRow.vala

```diff
@@ -0,0 +1,176 @@
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* General Public License for more details.
+*
+* You should have received a copy of the GNU General Public
+* License along with this program; if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*
+* Authored by: Marius Meisenzahl <mariusmeisenzahl@gmail.com>
+*/
+
+public class About.Widgets.FirmwareUpdateRow : Gtk.ListBoxRow {
+    public Device device { get; construct set; }
+
+    public signal void on_update_start ();
+    public signal void on_update_end ();
+
+    public FirmwareUpdateRow (Device device) {
+        Object (device: device);
+    }
+
+    construct {
+        var icon = new Gtk.Image.from_icon_name (device.icon, Gtk.IconSize.DND) {
+            pixel_size = 32
+        };
+
+        var device_name_label = new Gtk.Label (device.name) {
+            halign = Gtk.Align.START,
+            hexpand = true
+        };
+        device_name_label.get_style_context ().add_class (Granite.STYLE_CLASS_H3_LABEL);
+
+        var version_label = new Gtk.Label (device.latest_release.version) {
+            wrap = true,
+            xalign = 0
+        };
+
+        var grid = new Gtk.Grid () {
+            column_spacing = 12,
+            margin = 6
+        };
+        grid.attach (icon, 0, 0, 1, 2);
+        grid.attach (device_name_label, 1, 0);
+        grid.attach (version_label, 1, 1);
+
+        switch (device.latest_release.flag) {
+            case ReleaseFlag.IS_UPGRADE:
+                if (device.latest_release.version == device.version) {
+                    add_up_to_date_label (grid);
+                    break;
+                }
+
+                var update_button = new Gtk.Button.with_label (_("Update")) {
+                    valign = Gtk.Align.CENTER
+                };
+                update_button.clicked.connect (() => {
+                    on_update_start ();
+
+                    update.begin (device, device.latest_release, (obj, res) => {
+                        update.end (res);
+                        on_update_end ();
+                    });
+                });
+                grid.attach (update_button, 2, 0, 1, 2);
+                break;
+            default:
+                add_up_to_date_label (grid);
+                break;
+        }
+
+        add (grid);
+    }
+
+    private void add_up_to_date_label (Gtk.Grid grid) {
+        var update_to_date_label = new Gtk.Label (_("Up to date")) {
+            valign = Gtk.Align.CENTER
+        };
+        grid.attach (update_to_date_label, 2, 0, 1, 2);
+    }
+
+    private async void update (Device device, Release release) {
+        var path = yield FwupdManager.get_instance ().download_file (device, release.uri);
+
+        var details = yield FwupdManager.get_instance ().get_details (device, path);
```

Comment:
Again, is there some way you can switch this around so you can just do, `device.get_details (path)`. Are these details about the update or release? If so, we should probably rename the method to reflect that (i.e. `get_update_details` or `get_release_details`).

Reply:
These details are release details. So I renamed it to `get_release_details`.

This method calls the D-Bus interface. So I would rather not move it to the `Device` class.

Reply:
Yes, I see now. I was thinking about how we could try and make the Fwupd interface feel a bit more object orientated, but since the DBus interface hasn't really been written in that way, it just makes things more complicated. It looks slightly cleaner now we're not using a singleton anyway, so we can forget this for now!